File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        //prefix and suffix arrays
        int n = height.length;

        int[] prefix = new int[n];
        int[] suffix = new int[n];
        prefix[0] = height[0];
        suffix[n - 1] = height[n - 1];

        for(int i = 1 ; i < n ; i++){
            prefix[i] = Math.max(prefix[i - 1] , height[i]);
        }

        for(int i = n - 2  ; i >= 0 ; i--){
            suffix[i] = Math.max(suffix[i + 1] , height[i]);
        }

        int left = 0 ;
        int right = n - 1;
        int max_area = 0;
        int area = 0;

        while(left < right){
            area = Math.min(prefix[left] , suffix[right]) * (right - left);
            max_area = Math.max(area , max_area);
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return max_area;
    }
}

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        area = 0
        i = 0
        j = len(height) - 1
        while(i < j):
            if height[i] <= height[j]:
                result = height[i] * (j - i)
                i += 1
            else:
                result = height[j] * (j - i)
                j -= 1
            if result > area:
                area = result
        return area
            